Mathlib/Tactic/RelCongr/Core.lean

```diff
@@ -0,0 +1,479 @@
+  let .app (.app rel lhs) rhs ← withReducible g.getType'
+    | throwError "rel_congr failed, not a relation"
+  let some relName := rel.getAppFn.constName?
+    | throwError "rel_congr failed, relation head {rel} is not a constant"
+  let (some lhsHead, lhsArgs) := lhs.withApp fun e a => (e.constName?, a)
+    | if template.isNone then return (names, #[g])
+      throwError "rel_congr failed, {lhs} is not a constant"
+  let (some rhsHead, rhsArgs) := rhs.withApp fun e a => (e.constName?, a)
+    | if template.isNone then return (names, #[g])
+      throwError "rel_congr failed, {rhs} is not a constant"
+  -- B. If there is a template, check that it is of the form `tplHead _ ... _` and that
+  -- `tplHead = lhsHead = rhsHead`
+  let tplArgs ← if let some tpl := template then
+    let (some tplHead, tplArgs) := tpl.withApp fun e a => (e.constName?, a)
+      | throwError "rel_congr failed, {tpl} is not a constant"
+    unless tplHead == lhsHead && tplArgs.size == rhsArgs.size do
+      throwError "expected {tplHead}, got {lhsHead}\n{lhs}"
+    unless tplHead == rhsHead && tplArgs.size == rhsArgs.size do
+      throwError "expected {tplHead}, got {rhsHead}\n{rhs}"
+    -- and also build a array of `Expr` corresponding to the arguments `_ ... _` to `tplHead` in the
+    -- template (these will be used in recursive calls later), and a array of booleans according to
+    -- which of these contain `?_`
+    tplArgs.mapM fun tpl => do
+      let mctx ← getMCtx
+      let hasMVar := tpl.findMVar? fun mvarId =>
+        if let some mdecl := mctx.findDecl? mvarId then
+          mdecl.kind matches .syntheticOpaque
+        else
+          false
+      pure (some tpl, hasMVar.isSome)
+  -- A. If there is no template, check that `lhs = rhs`
+  else
+    unless lhsHead == rhsHead && lhsArgs.size == rhsArgs.size do
+      -- (if not, stop and report the existing goal)
+      return (names, #[g])
+    -- and also build a array of booleans according to which arguments `_ ... _` to the head
+    -- function differ between the LHS and RHS
+    (lhsArgs.zip rhsArgs).mapM fun (lhsArg, rhsArg) =>
+      return (none, !(← isDefEq lhsArg rhsArg))
+  -- Name the array of booleans `varyingArgs`: this records which arguments to the head function are
+  -- supposed to vary, according to the template (if there is one), and in the absence of a template
+  -- to record which arguments to the head function differ between the two sides of the goal.
+  let varyingArgs := tplArgs.map (·.2)
+  if varyingArgs.all not then
+    throwError "try refl"
+  let s ← saveState
+  let mut ex? := none
+  -- Look up the `@[rel_congr]` lemmas whose conclusion has the same relation and head function as
+  -- the goal.
+  for lem in (relCongrExt.getState (← getEnv)).findD (relName, lhsHead) #[] do
+    -- Filter further according to whether the boolean-array of varying/nonvarying arguments of such
+    -- a lemma matches `varyingArgs`.
+    if lem.varyingArgs == varyingArgs then
```

Comment:
It seems that my expectation is true, which is a design decision we have to be careful with (that if we have a `tt ~ tt` lemma, we also need the corresponding `tt ~ ff` and `ff ~ tt` lemmas).

Reply:
The initial plan was to have this automatically select the next best match and close the extra goals by refl, but @hrmacbeth argued that it should be fine since we don't currently have any cases where there are such holes in the coverage, and it is not a huge ask to keep it that way.

Reply:
Yes ... my opinion: it seems like the correct method to find the "next best" lemma would be rather heavyweight to implement (consider all tagged lemmas whose list-of-booleans is greater than or equal to the list-of-booleans represented by the goal, wrt to the partial order on lists of booleans, and select a lemma which is a minimal element of this set), so a lot of work, considering the current setting when there are only about 20 lemmas tagged anyway.

If the number of tagged lemmas starts to explode, it would certainly make sense to revisit this.

Reply:
As long as we are using equality though for the list-of-bools, we could use it as an additional hash key instead of looping over all of them.

Reply:
I think this is reasonable. It should be clearly documented though.